Enables E3SM - GCAM coupling

singhbalwinder opened the pull request:
Integrate the E3SM's Human Component (EHC) into the E3SM coupler framework, enabling dynamic coupling with GCAM, ELM, and EAM.

The EHC now optionally operates as a full E3SM component, exchanging CO2 emissions and land-use/cover data 
through the coupler. GCAM is called every 5 years to generate scenario data, which is processed and passed to 
ELM and EAM each model year. Terrestrial productivity feedback (NPP, HR) from ELM is averaged over 5 years and 
returned to GCAM.

Key updates include:

EHC-ELM feedback coupling managed by the coupler (no file-based transfer).
Implementation of scaling factors for GCAM yield and carbon density based on ELM NPP and HR fluxes.
Outlier handling and physical bounds of scaling factors for stability.
Annual land-use downscaling and conversion from GCAM to ELM via EHC-specific code, GLM, the LUT, and mksurfdata.
EHC to EAM coupling for annual CO2 emissions with spatial and vertical scaling.
Support for gridded EAM forcing inputs derived from GCAM and CMIP6 data.

This enhancement establishes a fully coupled human-Earth system workflow within E3SM, enabling consistent socioeconomic-climate feedbacks.

[BFB]

singhbalwinder commented:
This is a draft PR. I am still working on fixing some of the code and running tests. Please feel free to take a look at the PR.

aldivi commented on cime_config/allactive/config_compsets.xml:
@singhbalwinder @evasinha Have we tested these compsets? Are these used at all?

rljacob commented on components/eam/src/chemistry/utils/iac_coupled_fields.F90:
```diff
@@ -0,0 +1,385 @@
+module iac_coupled_fields
```
Shouldn't this file be in eam/src/utils ?  Its not used only by the chemistry.

singhbalwinder replied:
Moved file to `eam/src/utils/`.

aldivi commented on components/elm/src/data_types/ColumnDataType.F90:
@singhbalwinder @evasinha this may be contributing to the H0 NON-BFB for the seedC variable. We are not using the crop model or the seed C, but it is still output in the H0 file.

aldivi commented on components/elm/src/main/controlMod.F90:
@singhbalwinder I think there are a few errors in this file, particularly related to fates options. line numbers are for the new file on the right. Line 810 and lines 883-894 need to be reverted to what is in the original master file.
I also see an inconsistency in the master file with respect to lines  1172-1174 in the new file; These lines have been removed from the print function in the master, so they should either be removed from the new file or the if expression needs to include use.fates like the mpi broadcast in the previous function.

jenniferholm replied:
@glemieux - Alan was noting that there are some FATES lines in this PR that might need some review (rebasing to add in GCAM)

aldivi commented on components/elm/src/main/landunit_varcon.F90:
@singhbalwinder @evasinha more reinserted polygon tundra

aldivi commented on components/gcam/bld/namelist_files/namelist_defaults_gcam.xml:
@singhbalwinder line 73 <write_co2> we should probably change to false

singhbalwinder replied:
Done.

aldivi commented on components/gcam/src:
@singhbalwinder this will also be updated with the bug fixes

singhbalwinder commented on components/data_comps/datm/cime_config/config_component.xml:
Probably not. I will revert it.

Some changes have been added during the rebase because this branch has a long and very convoluted history.

singhbalwinder commented on components/data_comps/drof/cime_config/config_component.xml:
These also don't seem relevant. I will revert this whole file.

singhbalwinder commented on components/elm/bld/namelist_files/namelist_defaults.xml:
I will revert it.

singhbalwinder replied:
We need this change. My IAC simulations were not able to find `flanduse_timeseries` without removing `sim_year_range="1850-2015"` from 
```
<flanduse_timeseries hgrid="0.9x1.25"      sim_year_range="1850-2015"
 use_crop=".false." >lnd/clm2/surfdata_map/landuse.timeseries_0.9x1.25_hist_simyr1850-2015_c180306.nc</flanduse_timeseries>
```
@evasinha @aldivi

singhbalwinder replied:
My simulations are EAM-ELM-GCAM (or an F case with GCAM), and they look for this file during the build namelist process.

singhbalwinder replied:
I have removed the sim_year, and the simulation worked fine on my end.

singhbalwinder commented:
I have reverted the following files based on the feedback:
```
cime_config/allactive/config_compsets.xml
components/data_comps/datm/cime_config/config_component.xml
components/data_comps/drof/cime_config/config_component.xml
components/eam/cime_config/buildnml
components/elm/bld/namelist_files/namelist_defaults.xml
components/elm/src/main/initGridCellsMod.F90
```

bpbond commented on components/elm/src/cpl/lnd_import_export.F90:
```diff
@@ -1489,6 +1511,14 @@ subroutine lnd_export( bounds, lnd2atm_vars, lnd2glc_vars, l2x)
           end do
        end if
 
+       ! iac coupling - need an iac_active logical somewhere
```
Confusing comment

singhbalwinder replied:
@aldivi : If I am understanding it correctly, we should remove this comment.

singhbalwinder replied:
Removed the comment.

bpbond commented on components/elm/src/dyn_subgrid/dynSubgridAdjustmentsMod.F90:
```diff
@@ -876,10 +885,12 @@ subroutine dyn_col_ns_Adjustments(bounds, clump_index, column_state_updater, col
-           col_ns%dyn_nbal_adjustments(begc:endc) + &
-           adjustment_one_level(begc:endc) * dzsoi_decomp(j)
-
+       ! TRS NaN trap, for debugging
```
1. The comments say "for debugging"...are these NaN traps still needed? If not, remove; If yes, remove the "for debugging" language
2. Remove Tim Shippert's initials; not needed and potentially confusing

singhbalwinder replied:
I have replaced these checks with simple if statements, where the code would exit if it detects a NaN.  That is, we won't skip any calculations based on NaN detection; the code will exit and let us know that it has detected NaNs. As we all know, NaNs floating around in a simulation can cause all kinds of headaches!

bpbond commented on components/elm/src/dyn_subgrid/dynSubgridControlMod.F90:
```diff
@@ -155,6 +155,19 @@ subroutine read_namelist( NLFilename )
     endif
 
+    ! the pft, crop, and harvest flags and landuse timeseries should be false or unset
+    !    if the ehc is active
```
See comment above about standardizing on "EHC" or "IAC"

bpbond reviewed (changes requested):
Woot! Thanks @singhbalwinder

aldivi commented on components/eam/bld/namelist_files/use_cases/SSP245_eam_EHC.xml:
@singhbalwinder @evasinha Do these use cases (and the elm ones) include the correct files? I know that some were recently updated.

evasinha replied:
@aldivi Yes, the `bndtvghg` files have been modified:

For the SSP245 use case:
```
<bndtvghg>atm/cam/ggas/GHG_GCAM_SSP245_Annual_Global_2015-2102_c20240712.nc</bndtvghg>
```

and for the SSP370 use case:
```
<bndtvghg>atm/cam/ggas/GHG_GCAM_SSP3_Annual_Global_2015-2102_c20240819.nc</bndtvghg>
```

singhbalwinder replied:
@evasinha : Could you please verify if these changes have been incorporated correctly? Let me know if we are missing these. Thanks.

singhbalwinder replied:
Thanks, @evasinha. I have updated the use cases. Could you please check if everything looks good for the use case files? Thanks.

aldivi commented on components/eam/src/cpl/cam_cpl_indices.F90:
```diff
@@ -88,13 +89,25 @@ module cam_cpl_indices
   integer :: index_x2a_Sx_u10          ! 10m wind
   integer :: index_x2a_Sx_u10withgusts ! 10m wind with gusts
 
+  ! Iac co2 fields - sfc, two air levels, monthly
```
@bpbond Yes, I agree. My intent with this naming is to eliminate the giac repo and bring the code into the E3SM repo component called 'EHC.' This  would eliminate the confusing double naming of 'gcam.' The 'iac' variables would be renamed during this proposed change. There are a lot of them.

singhbalwinder commented on components/eam/src/physics/cam/polar_avg.F90:
```diff
@@ -179,6 +180,9 @@ subroutine polar_average3d(nlev, field)
      end do
+
+n_pole=0
+s_pole=0
```
@aldivi and @evasinha: Do we need these changes?

aldivi replied:
@singhbalwinder I don't know. maybe they were removed from v2 because they were not necessary? I am not familiar with this code

singhbalwinder commented on components/elm/src/main/subgridAveMod.F90:
```diff
@@ -1070,6 +1071,10 @@ subroutine c2g_1d(bounds, carr, garr, c2l_scale_type, l2g_scale_type)
     sumwt(bounds%begg : bounds%endg) = 0._r8
     do c = bounds%begc,bounds%endc
+       ! TRS - nans during initialization cause problems with debugger.
+       if (isnan(carr(c))) then
```
We should exit if it is NaN

rljacob commented on cime_config/machines/cmake_macros/intel.cmake:
```diff
@@ -1,24 +1,24 @@
-string(APPEND CMAKE_Fortran_FORMAT_FREE_FLAG " -free")
+string(APPEND CMAKE_Fortran_FLAGS " -convert big_endian -assume byterecl -ftz -traceback -assume realloc_lhs -fp-model source -mcmodel=medium  -fPIC")
+string(APPEND CMAKE_Fortran_FORMAT_FIXED_FLAG " -fixed -132  -mcmodel=medium  -fPIC")
+string(APPEND CMAKE_Fortran_FORMAT_FREE_FLAG " -free  -mcmodel=medium  -fPIC")
```
Have you verified that these flags have no effect on answers or performance on cases without GCAM?  On every machine where intel is the main compiler?

singhbalwinder replied:
I have never done any testing, but these flags may affect the model's performance. As mentioned above, we should either add these flags only in the specific machine files that require them (rather than in the general Intel file), or identify the F90 files that need them and fix those if possible.

rljacob commented on components/eam/src/chemistry/utils/iac_coupled_fields.F90:
```diff
@@ -0,0 +1,385 @@
+
+    integer            :: idx
+
+    ! FIXMEB: Hardwire it true for now - I don't know if this is an infodata or namelist
```
iac_present should be obtained through the infobuffer in the import_mct call using a call to seq_infodata_GetData.  Then copy it to a module variable to use in the rest of EAM.

aldivi commented on components/elm/src/main/elm_varctl.F90:
```diff
@@ -373,7 +387,7 @@ module elm_varctl
   logical, public :: use_crop            = .false.
   logical, public :: use_snicar_frc      = .false.
-  logical, public :: use_snicar_ad       = .false.
+  logical, public :: use_snicar_ad       = .true.
```
@singhbalwinder @evasinha It appears to be true in the current master and false in the 3.0.2 tag. So maybe we keep it true to be consistent with the current master? it appears to be a multilayer snow scheme.

aldivi commented on components/gcam/cime_config/user_nl_gcam:
@beharrop I have put off changing names like this until we can do a more complete change. For example, the giac repo should be merged into the E3SM repo as the ehc component (not gcam). There is a lot of code labelled 'iac' that should also change. It may not be a huge lift to change this particular filename now, but then it wouldn't match the 'gcam' component name.

singhbalwinder commented:
I have updated the gcam/giac submodules and also rebased this branch over the current master.

singhbalwinder commented:
I tried `ERS_Ld5_P4.ne4pg2_oQU480.F2010.compy_intel.eam-wcprod_F2010` on Compy and it is working fine for me. It may be a compiler issue. I will test with the GNU compiler.

bishtgautam commented on components/elm/src/main/iac2lndMod.F90:
```diff
@@ -0,0 +1,327 @@
+
+
+  !---------------------------------------------------------------------
+  subroutine iac_rpointer_write(rdate)
```
Deleting code to longer used is even better :)

singhbalwinder replied:
@aldivi: Should I also delete [these](https://github.com/E3SM-Project/E3SM/blob/5052531cdc1a74182393da32c5eaf9acbf5f7d89/components/elm/src/main/elm_driver.F90#L1559) calls for `iac_rpointer_write`?

aldivi replied:
@singhbalwinder yes, this and the calls should be deleted as per the bug fix commits.

singhbalwinder commented on components/eam/src/physics/cam/physpkg.F90:
```diff
@@ -1804,10 +1813,18 @@ subroutine tphysac (ztodt,   cam_in,  &
+      call physics_update(state, ptend, ztodt, tend)
+      ! Compute diagnostics (supply optional iac_co2_name to get_carbon_air_fluxes for iac)
+      call get_carbon_air_fluxes(state, pbuf, ztodt, iac_co2_name)
+    else if(co2_readFlux_aircraft) then
```
That is a good catch. This should be a simple `else` to retain the default behavior is `iac_present` is `.false.`

singhbalwinder commented on components/elm/src/cpl/lnd_comp_mct.F90:
```diff
@@ -357,12 +363,12 @@ subroutine lnd_init_mct( EClock, cdata_l, x2l_l, l2x_l, NLFilename )
 #ifdef HAVE_MOAB
 !     Also send data through the MOAB path in driver-moab
-      call lnd_export_moab(EClock, bounds, lnd2atm_vars, lnd2glc_vars) ! it is private here
+      call lnd_export_moab(EClock, bounds, lnd2atm_vars, lnd2glc_vars, lnd2iac_vars) ! it is private here
```
Thanks. I will delete this line.

bishtgautam reviewed (commented):
Just a thought: While a low-res configuration with active GCAM might not be available just yet for performing a multi-long year needed for testing GCAM coupling, how about adding a 5-day ERS test for the resolution that currently works with GCAM to at least make sure new PRs (especially related to module updates) don't break this new capability?

singhbalwinder commented on components/elm/src/cpl/elm_cpl_indices.F90:
```diff
@@ -347,6 +363,38 @@ subroutine elm_cpl_indices_set( )
+    ! This is called before the pft number is set 
+    ! and these are indexed from zero on this elm side 
+    ! KVC: this has a different value than in iac. Need to align. Doing manually now
+    iac_npft = 17
```
@bishtgautam suggested adding a check here. Which variables should I check and compare? @aldivi

aldivi replied:
@singhbalwinder @bishtgautam the elm variables mxpft_nc and numharvest have not been set yet at this point. They are in elm_varpar.F90, with elm_varpar_init() called in initialize1, which is called after elm_cpl_indices_set. So the earliest check could be in elm_varpar_init: use only iac_npft and iac_nharvest from the elm_cpl_indices module and check them agains mxpft_nc (<=) and numharvest (equality).
The reason that iac_npft does not align with the active numpft variable in elm is because numpft does not include the barepft. Maybe a better check than above would be checking equality between iac_npft and numpft+1

singhbalwinder replied:
I have added two checks for these [here](https://github.com/E3SM-Project/E3SM/pull/7536/files#diff-1f858edd36e7af8537c3e446dc34b648e3fcbbfe1e579638d6303e51bb758190R256). Do they look appropriate?

aldivi replied:
@singhbalwinder I just don't think they should be defined in two places; they should be defined/set only one time, in one place. If you can load them from elm_varpar into elm_cpl_indices, then that should work. Just note that the cpl init happens before the varpar init. I think Gautam suggested these checks to make sure that these ehc variables were consistent with the elm variables. It is true that they are hardcoded so they should be fine, although if the crop model is on the numpft check will fail, but I think we check to make sure that the crop model is not on when the ehc is active. The checks would be useful if we make the ehc variables namelist items for elm, but so far we are not doing this.

singhbalwinder replied:
@aldivi:
>  I just don't think they should be defined in two places; they should be defined/set only one time, in one place. 

In the current version, the iac variables (`iac_npft` and `iac_nharvest`) are set only at one place (`elm_varpar.F90`). They are then being used at other places. I don't see them defined anywhere else. Please let me know if I misunderstood your comment.

> If you can load them from elm_varpar into elm_cpl_indices, then that should work. 

I think that is what I am doing currently. Again, please let me know if I misunderstood your comment.

>although if the crop model is on the numpft check will fail, but I think we check to make sure that the crop model is not on when the ehc is active.

Yes, you have added a check for the crop model (see [here](https://github.com/E3SM-Project/E3SM/blob/021b17be6a4f8812379bda1689c431d666a393c0/components/elm/src/dyn_subgrid/dynSubgridControlMod.F90#L161))

What changes do you suggest I should make to these checks?

singhbalwinder commented on components/elm/src/dyn_subgrid/dynSubgridDriverMod.F90:
```diff
@@ -368,4 +406,169 @@ subroutine dynSubgrid_wrapup_weight_changes(bounds_clump, glc2lnd_vars)
+       t = veg_pp%topounit(p)
+       g = veg_pp%gridcell(p)
+       if (col_pp%wtgcell(c) .eq. 0._r8) then
+          !write(iulog,*) subname//'setting wtcol to zero' // &
+           !              'for p, c, l, g ', p,c,l,g
```
Done

singhbalwinder commented on components/elm/src/dyn_subgrid/dynSubgridDriverMod.F90:
```diff
@@ -368,4 +406,169 @@ subroutine dynSubgrid_wrapup_weight_changes(bounds_clump, glc2lnd_vars)
+       veg_pp%wtgcell(p)    = veg_pp%wtcol(p) * col_pp%wtgcell(c)
+       ! get the sums for normalizing below
+       sumwtcol(c) = sumwtcol(c) + veg_pp%wtcol(p)
+       !sumwtlunit(l) = sumwtlunit(l) + veg_pp%wtlunit(p)
+       !sumwttopounit(t) = sumwttopounit(t) + veg_pp%wttopounit(p)
+       !sumwtgcell(g) = sumwtgcell(g) + veg_pp%wtgcell(p)
```
Done

singhbalwinder commented on components/elm/src/dyn_subgrid/dynSubgridDriverMod.F90:
```diff
@@ -368,4 +406,169 @@ subroutine dynSubgrid_wrapup_weight_changes(bounds_clump, glc2lnd_vars)
+                veg_pp%wtgcell(p)    = veg_pp%wtcol(p) * col_pp%wtgcell(c)
+
+
+                ! get the higher order sums
+                !l = veg_pp%landunit(p)
+                !t = veg_pp%topounit(p)
+                !g = veg_pp%gridcell(p)
+                !sumwtlunit(l) = sumwtlunit(l) + veg_pp%wtlunit(p)
+                !sumwttopounit(t) = sumwttopounit(t) + veg_pp%wttopounit(p)
+                !sumwtgcell(g) = sumwtgcell(g) + veg_pp%wtgcell(p)
```
Done.

singhbalwinder commented on components/elm/src/main/iac2lndMod.F90:
```diff
@@ -0,0 +1,327 @@
+
+!avd
+!write(iulog,*) subname, 'begg,endg=', begg,endg
+!write(iulog,*) subname, 'begp,endp=', begp,endp
```
Done.

singhbalwinder commented on components/elm/src/main/iac2lndMod.F90:
```diff
@@ -0,0 +1,327 @@
+       ierr = nf90_put_var(ncid,cid_varid,cell_ids)
+
+       ierr = nf90_close(ncid)
+end if
```
Done.

singhbalwinder commented on components/elm/src/dyn_subgrid/dynSubgridDriverMod.F90:
```diff
@@ -368,4 +406,169 @@ subroutine dynSubgrid_wrapup_weight_changes(bounds_clump, glc2lnd_vars)
+    do c = bounds%begc,bounds%endc
+       weights_equal_1 = (abs(sumwtcol(c) - 1._r8) <= tolerance)
+       if (.not.weights_equal_1) then
+          !write(iulog,*) subname//'weights not equal 1, sumwtcol = ', &
+                !           sumwtcol(c), 'c= ', c
```
Done.

singhbalwinder commented on components/elm/src/main/elm_varcon.F90:
```diff
@@ -72,6 +72,10 @@ module elm_varcon
   real(r8) :: tlsai_crit = 2.0_r8                           ! critical value of elai+esai for which aerodynamic parameters are maximum
   real(r8) :: watmin = 0.01_r8                              ! minimum soil moisture (mm)
 
+  real(r8), parameter :: mm_epsilon = 0.622_r8              ! Molar mass ratio (water:dry air)
```
I am not sure. This the not a change this PR brought in (as far as I know). @bishtgautam @aldivi do you guys know about this?

aldivi commented on components/eam/src/chemistry/utils/iac_coupled_fields.F90:
```diff
@@ -0,0 +1,385 @@
+    type(physics_buffer_desc), pointer :: pbuf_chnk(:)
+
+    character(len=cxx)  :: err_str
+    integer, parameter :: HIGH_LAYER = 11000 !Height at which fco2_high should be released [m]
```
@singhbalwinder not really; it was determined by the baseline CEDS aircraft co2 data, and then processed this way. if not hardwired it would come in through an eam namelist item.

thorntonpe commented:
Thanks for those details, @aldivi .

aldivi commented:
@thorntonpe @singhbalwinder I think I can address your topounit question also. We have not tested E3SM-GCAM with multiple topounits, and we don't support it. The only check I have found related to this is on lines 487-495 in E3SM/components/elm/src/main/subgridRestMod.F90 to ensure that the correct weights are read in. The initial files for starting E3SM-GCAM in 2015 will not have multiple topounits. It may be worth checking that the variable max_topounits==1 when iac_present is true, in initialize1() in .../elm/src/main/elm_initializeMod.F90. if not, the model can abort.

singhbalwinder commented on components/elm/src/main/lnd2iacMod.F90:
Done.

singhbalwinder commented on components/elm/src/main/histFileMod.F90:
Reverted this to the original file.

singhbalwinder commented:
@aldivi : I have added a check for topounits. It looks like the default `max_topounit` is zero (not 1). I added a check for zero. Please let me know if it is unexpected. I simulated with EHC turned on, and the `max_topounits` in my simulation was 0.

aldivi commented:
@singhbalwinder the working value for max_topounits if they do not exist is 1. This is set by topounit_varcon_init, which is called directly after the check you inserted in elm initialize1. So move the check to just after the topounit call, and check that max_topounits=1 when EHC is present

bpbond commented:
>have your requested changes been made?

@rljacob Yes, thanks

rljacob commented:
@singhbalwinder if you haven't, please make sure e3sm_developer will pass on this branch.

singhbalwinder commented on components/elm/src/main/controlMod.F90:
```diff
@@ -466,6 +466,11 @@ subroutine control_init( )
+             call endrun(msg=' ERROR: use_lch4 (methane) and use_fates cannot both be set to true.'//&
+                   errMsg(__FILE__, __LINE__))
+          end if
+
```
Just so you know, the test passes if I remove this if condition.

evasinha replied:
@singhbalwinder This if condition was added in this [commit](https://github.com/E3SM-Project/E3SM/commit/805450b604c60dfb9761fb9b33043a9c53f6b358#diff-503880584316cf787a6f4ea9a66ce27dbb98a910610a649680684903d935e95a). @aldivi  think it can be removed.

singhbalwinder commented on components/elm/src/main/subgridAveMod.F90:
```diff
@@ -1009,6 +1010,11 @@ subroutine c2g_1d(bounds, carr, garr, c2l_scale_type, l2g_scale_type)
+       if (isnan(carr(c))) then
+          call endrun(msg='carr(c) is NaN '//&
+                errMsg(__FILE__, __LINE__))
+       endif
```
`SMS.r05_r05.IELM.chrysalis_intel.elm-topounit_im2` is failing because of this NaN check. We should not have NaNs floating around in a simulation. I will run some more tests to know more.

evasinha commented on components/elm/src/main/controlMod.F90:
```diff
@@ -518,6 +523,11 @@ subroutine control_init( )
+       if (.not. use_crop .and. irrigate) then
+          call endrun(msg=' ERROR: irrigate = .true. requires CROP model active.'//&
+            errMsg(__FILE__, __LINE__))
+       end if
```
@singhbalwinder This if condition was added in this [commit](https://github.com/E3SM-Project/E3SM/commit/805450b604c60dfb9761fb9b33043a9c53f6b358#diff-503880584316cf787a6f4ea9a66ce27dbb98a910610a649680684903d935e95a). @aldivi  think it can be removed.

singhbalwinder replied:
I will remove this too.

aldivi replied:
@singhbalwinder @evasinha I think if the ch4 and crop&irrigate checks are not in the master then they can definitely be removed. I don't recall adding them, and they could have been legacy checks.

rljacob commented on driver-mct/main/cime_comp_mod.F90:
```diff
@@ -1777,7 +1791,7 @@ subroutine cime_init()
        if (rof_prognostic) lnd_c2_rof = .true.
        if (glc_prognostic) lnd_c2_glc = .true.
-       if (iac_prognostic) lnd_c2_iac = .true.
+       if (iac_present) lnd_c2_iac = .true.
```
Follow the pattern and use "iac_prognostic" here.

singhbalwinder commented on driver-mct/main/cime_comp_mod.F90:
```diff
@@ -4744,12 +4789,21 @@ subroutine cime_run_calc_budgets1(in_cplrun)
        if (lnd_present) then
           call seq_diag_lnd_mct(lnd(ens1), fractions_lx(ens1), infodata, do_l2x=.true., do_x2l=.true.)
+          if (do_bgc_budgets) then
+             call seq_diagBGC_lnd_mct(lnd(ens1), fractions_lx(ens1), infodata, do_l2x=.true., do_x2l=.true.)
```
@aldivi : Do you know why these calls are added?
